### PR TITLE
ResourceTags Migration

### PR DESCRIPTION
**Pull request by skellamp**
TODO

- [x] Test deduplication service using Solr FallbackLoader.

### CODE REVIEW

File: module/VuFind/src/VuFind/ChannelProvider/ListItemsFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/ChannelProvider/ListItemsFactory.php:74]: Would it be better to use the fully-qualified class name here instead of the alias? That's what we seem to be doing most of the time, and it's good to be consistent.

File: module/VuFind/src/VuFind/Db/Entity/ResourceTags.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Entity/ResourceTags.php:133]: Worth adding `: int` type here?

File: module/VuFind/src/VuFind/Db/Entity/Tags.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Entity/Tags.php:76]: Worth adding type here as well?

File: module/VuFind/src/VuFind/Db/Entity/UserList.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Entity/UserList.php:122]: ...and type here too?

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Resource.php:210]: Is this method really just returning an array of IDs instead of an array of objects? If so, I wonder if it would make sense to rename it to `getResourceIDsForTag` to clarify the nature of the return value.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/TagService.php:455]: Typo: reosurce

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Resource.php:249]: Since we're breaking the pattern of the loop, maybe it's better to do the database service stuff outside of the loop instead of introducing a special case here. It also looks like there is no `update` method on the tag service -- I think you mean to be calling `updateResource` but maybe a better name for that method would be `changeResourceId`.

Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Table/Resource.php:249]: It was left as tableObjects, after this I will migrate rest of the things and change it to serviceObjects under a loop.

**Comment by skellamp** [comment]
> Thanks, @skellamp! I haven't tried this yet, but I've left a few comments based on an initial readthrough of the code. I didn't give the TagService itself a close look yet since I like to review that in comparison to the code that is being removed, and I see that you haven't removed the old code yet. Is there a reason you are still retaining it -- i.e. is the legacy code still in use somewhere -- or is that something else you can do as part of the next batch of updates here?

Once the testing is done, will perform the cleanup operation.

File: module/VuFind/src/VuFind/Db/Table/Tags.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Tags.php:526]: Did you mean to delete this? I think it's still currently being used by VuFind\Db\Table\ResourceTags. Is there something else you meant to delete instead, maybe from the Doctrine code rather than the Laminas code?

**Review by demiankatz — CHANGES_REQUESTED** [review]
I hope to do more testing soon, but I noticed the issue below when reviewing your most recent commit.

File: module/VuFind/src/VuFind/Db/Table/Tags.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Table/Tags.php:526]: It was migrated to TagService and no longer useful. It was removed intentionally.

Comment (demiankatz) [reply module/VuFind/src/VuFind/Db/Table/Tags.php:526]: Ahh, sorry, I got confused because ResourceTags::processDestroyLinks hasn't been deleted yet. But of course that's what we just discussed -- I will work on deleting that and other obsolete methods as part of my testing process. :-)

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/TagService.php:243]: When working on my pull request to fix existing bugs in this method, I discovered that selecting tagcnt here and using it in the HAVING clause below does not work with PostgreSQL. There is no reason to select it anyway, so I suggest removing this clause here, and changing the HAVING below to directly embed `COUNT(DISTINCT(rt.tag))` in place of the name `tagcnt`.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/TagService.php:247]: The original code includes this, which doesn't seem to have been ported over:

```
            // Restrict to tags by list owner
            $select->where->and->equalTo(
                'resource_tags.user_id',
                new Expression('l.user_id')
            );
```

I think you could just add `AND rt.user_id = l.user_id` or something similar to cover this situation.

I can't actually think of a strong reason why this is necessary -- by the nature of the system, users can only add tags to their own lists at present, and if we added shared lists in the future, we probably would NOT want this restriction... so maybe it's fine to leave it out... but I just wanted to check and see if you had already thought this through, or if it had been dropped by accident.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/TagService.php:254]: Since this is a hard-coded value rather than a user-provided variable, would it be better to hard-code the value into the DQL instead of providing it as an additional parameter? I don't mind this approach if you think it best, but all other things being equal, the hard-coded approach would shave off a line of code.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Service/TagService.php:263]: The original code is subtly different from this -- it does all the lowercasing in the database, instead of doing some in the database and some in PHP. I think it is better if we can continue to do things that way, because sometimes different regions or locales may have different lowercasing rules, and this could result in inconsistent behavior if the database and PHP apply different algorithms.

I'm not sure what the easiest way to do this would be, but I guess a crude but effective approach might be something like:

```
$tagClauses = [];
foreach ($tag as $i => $currentTag) {
    $tagPlaceholder = 'tag' . $i;
    $tagClauses[] = 'LOWER(t.tag) = LOWER(:' . $tagPlaceholder . ')';
    $parameters[$tagPlaceholder] = $currentTag;
}
$dql .= 'AND (' . implode(' OR ', $tagClauses) . ')';
```

...maybe Doctrine offers some clever way to do this in less code, but if not, I think this is okay.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@skellamp, I just did a review entirely focused on getListsForTag(), since my fixes to that method have just been merged into dev. I have a few suggested changes. I'll finish up my hands-on testing once you've had a chance to act on (or ignore, if appropriate) these suggestions.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Service/TagService.php:247]: That's the exact reason why I removed it, I just added it back.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Service/TagService.php:263]: I couldn't find a better way of doing it in Doctrine. I was surprised to know that case, I did some reading and got to know that the conversions often go wrong. The only concern I have is it might introduce a performance bottleneck and it's good to add a index over the tags in Lowercase.

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Resource.php:265]: This needs to be changed back to resourcetags, doesn't it? There's no deduplicate method in the tagService here -- I think we discussed doing this work in a separate PR after merging this one. Am I remembering correctly?

**Review by demiankatz — CHANGES_REQUESTED** [review]

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Resource.php:257]: Why are you removing this line? I think we do want to delete the old resource after associated rows have been migrated to the new resource.

File: module/VuFind/src/VuFind/Db/Table/Tags.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/Tags.php:572]: The Laminas ResourceTags table object is still being used in the line above. Did you leave that intentionally? If so, I have no objection to merging the current work first, and then doing a second PR to get rid of more ResourceTags references... but just mentioning in case it was somehow missed.

**Review by demiankatz — CHANGES_REQUESTED** [review]
A couple more things...

I think we're very close to being ready to merge this -- once these last few questions are answered/addressed, I should be able to finish up and merge, unless you want to keep this open in order to append further work.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Db/Service/TagService.php]: Yes, it's probably worth removing in future, but best to do that after migration so we don't confuse ourselves if there's some factor we're overlooking.

File: module/VuFind/src/VuFind/Db/Service/TagService.php
Comment (demiankatz) [reply module/VuFind/src/VuFind/Db/Service/TagService.php]: I think it's unlikely in practice to be run over a very large number of values, but if that use case comes up, you're right that an index might help.

I've found over the years that some of the most frustrating and challenging bugs come from issues related to character encoding and representation of time -- both things that seem like they should be simple, yet turn out not to be!

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Table/Resource.php:265]: That's right, working on it probably will raise a PR by eod.

File: module/VuFind/src/VuFind/Db/Table/Resource.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Table/Resource.php:257]: Sorry, that's a mistake, I just found it, working out of doctrine branch will fix this in PR on deduplication.

File: module/VuFind/src/VuFind/Db/Table/Tags.php
Comment (skellamp) [reply module/VuFind/src/VuFind/Db/Table/Tags.php:572]: Sorry, somehow I missed it, but we can move that along with Tags as it has dependency with thin Tags. Please merge the current work.

**Comment by skellamp** [comment]
> @skellamp, I tried to use the Solr FallbackLoader to test the resource merging functionality, and it's not quite working -- I think the problem is that we missed a related detail when we migrated comment logic earlier, and we need to fix it now.
> 
> Here's how to set up the test scenario:
> 
> 1.) In searches.ini, uncomment the `fallback_id_field` setting and set it to `ctrlnum`. This will put VuFind into a mode where, if it tries to do a lookup by ID and can't find a match in the regular ID field, it will fall back and check the ctrlnum field... and if it finds a match, it will then merge any existing data associated with the ctrlnum value into the associated id value.
> 
> 2.) Go to http://localhost/vufind_test/Record/vtls000000329 in a standard test environment, create an account, and save the record to your favorites. You can also add comments or other data if you wish to make the test case more complex.
> 
> 3.) In the database, run `update resource set record_id='(IeDuNL)1048';` -- this will change the record you just created so that its ID matches the ctrlnum value for the record instead of the id value, which will let us simulate a record ID change.
> 
> 4.) Now go to http://localhost/vufind_test/Record/(IeDuNL)1048 -- you should see exactly the same record that you saw in the other link.
> 
> 5.) Now try to save the record again -- this will trigger a resource merge, and you'll get an error about the comments table object not existing. You can probably incorporate the updating of the comments table into your existing updateResource method... but maybe that method should be moved out of the TagService and into the ResourceService -- and renamed to updateResourceID or mergeResourceIDs for clarity.
> 
> I hope this helps move things forward a little more, but please let me know if you run into trouble or if any of this is unclear. :-)

For record a detailed test scenario is describe here
https://vufind.org/wiki/development:testing:manual_testing#merging_record_data_when_ids_change

**Review by demiankatz — APPROVED** [review]
Thanks, @skellamp, this is looking great and all tests are passing.

Please feel free to open your next PR whenever you're ready for me to begin review.

**Pull request merged**
